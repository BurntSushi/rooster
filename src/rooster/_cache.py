import contextlib
import hashlib
import json

import hishel
import httpx
from httpcore import Request, Response


class GraphQLCacheController(hishel.Controller):
    def __init__(self):
        super().__init__(allow_heuristics=True)
        # Hack support for POST requests as hishel does not allow caching of them
        self._cacheable_methods = ["POST"]

    def is_cachable(self, request: Request, response: Response) -> bool:
        # Since GraphQL always returns a 200, we check the response body for errors
        error_in_response_body = False
        try:
            error_in_response_body = json.loads(response.read()).get("errors")
        except Exception:
            pass  # The response cannot be parsed as JSON
        return super().is_cachable(request, response) and not error_in_response_body


@contextlib.contextmanager
def graphql_http_client():
    """
    A HTTP client with support for caching GraphQL requests.

    The cache will be stored at `$PWD/.cache`.
    """
    with httpx.Client(
        transport=hishel.CacheTransport(
            transport=httpx.HTTPTransport(),
            controller=GraphQLCacheController(),
        ),
        timeout=None,
    ) as client:
        yield client


def _generate_key_patch(request: Request) -> str:
    """
    Patches Hishel's cache key generation to include the request body for GraphQL requests
    which are always hitting the same endpoint.
    """
    base_key = _generate_key(request)
    hash = hashlib.new("sha256")
    hash.update(base_key.encode())
    for item in request.stream:
        hash.update(item)
    patched_key = hash.hexdigest()
    return patched_key


# Eek! Hishel doesn't let you change this so I patch it for now instead of implementing
# the whole transport again
_generate_key = hishel._sync._transports.generate_key
hishel._sync._transports.generate_key = _generate_key_patch
