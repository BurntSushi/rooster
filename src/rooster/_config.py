from pathlib import Path

import pydantic
import tomllib


class Config(pydantic.BaseModel):
    major_labels: frozenset[str] = frozenset(["breaking"])
    minor_labels: frozenset[str] = frozenset(["feature"])
    patch_labels: frozenset[str] = frozenset(["fix"])

    changelog_sections: dict[str, str] = {
        "breaking": "Breaking changes",
        "feature": "New features",
        "fix": "Bug fixes",
    }
    changelog_ignore_labels: frozenset[str] = frozenset([])
    changelog_ignore_authors: frozenset[str] = frozenset(["dependabot"])

    change_template: str = "- {pull_request.title} (#{pull_request.number})"

    @pydantic.validator("changelog_sections", always=True)
    def require_unknown_key(cls, value):
        value.setdefault("__unknown__", "Other changes")
        return value


def get_config(repo: Path) -> Config:
    pyproject_path = repo / "pyproject.toml"
    if not pyproject_path.exists():
        return Config()
    pyproject = tomllib.loads(pyproject_path.read_text())
    section = pyproject.get("tool", {}).get("rooster", {})
    return Config(**section)
