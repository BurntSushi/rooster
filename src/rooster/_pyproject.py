from pathlib import Path

import tomllib
from packaging.version import Version


def update_pyproject_version(path: Path, version: Version) -> None:
    contents = path.read_text()
    parsed = tomllib.loads(contents)
    old_version = parsed["project"]["version"]
    contents = contents.replace(f'version = "{old_version}"', f'version = "{version}"')
    path.write_text(contents)
