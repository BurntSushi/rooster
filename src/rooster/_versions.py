"""
Utilities for working with version numbers.
"""

from enum import Enum
from pathlib import Path

from packaging.version import InvalidVersion, Version

from rooster._git import get_tags


class BumpType(Enum):
    major = "major"
    minor = "minor"
    patch = "patch"


def versions_from_git_tags(repo: Path) -> list[Version]:
    """
    Get versions of the project from git tags.
    """
    tags = get_tags(repo)
    versions = parse_versions(tags)
    return versions


def parse_versions(version_strings: list[str]) -> list[Version]:
    """
    Parse version strings typed, validated objects.

    Invalid versions will be silently ignored.
    """
    versions = []
    for version in version_strings:
        try:
            versions.append(Version(version))
        except InvalidVersion:
            # Ignore tags that are not valid versions
            pass
    return versions


def get_latest_version(versions: list[Version]) -> Version | None:
    """
    Get the newest version from a collection of versions.
    """
    if not versions:
        return None
    return sorted(versions, reverse=True)[0]


def get_previous_version(versions: list[Version], version: Version) -> Version | None:
    """
    Get the version preceding a given version from a collection of versions.
    """
    if version not in versions:
        # Ensure the given version is included (but do not modify the given list)
        versions = [version] + versions
    versions = sorted(versions)
    index = versions.index(version) - 1
    if index < 0:
        return None
    return versions[index]


def bump_version(version: Version, bump_type: BumpType) -> Version:
    """
    Create a new version from a preceding one, increasing the given component.
    """
    # Pull the release section from the version and increment the appropriate number
    release = list(version.release)

    match bump_type:
        case BumpType.patch:
            release[2] += 1
        case BumpType.minor:
            release[1] += 1
            release[2] = 0
        case BumpType.major:
            release[0] += 1
            # Reset minor and patch versions
            release[1] = 0
            release[2] = 0
        case _:
            raise ValueError(f"Invalid bump type: {bump_type}")

    # Build a new  version string (`Version` is immutable)
    parts = []
    if version.epoch != 0:
        parts.append(f"{version.epoch}!")

    # Update release segment
    parts.append(".".join(str(x) for x in release))

    # We do not include other sections like dev/local/post since we are publishing
    # a new version. We could allow doing so in a separate function but then we
    # would need to construct the object again.
    return Version("".join(parts))
