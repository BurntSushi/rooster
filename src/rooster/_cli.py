from pathlib import Path

import tomllib
import typer

from rooster._changelog import (
    add_or_update_entry,
    entry_to_standalone,
    extract_entry,
    generate_changelog,
    get_versions_from_changelog,
)
from rooster._config import get_config
from rooster._git import get_commits_between, get_remote_url
from rooster._github import get_pull_requests_for_commits, get_release, parse_remote_url
from rooster._versions import (
    Version,
    bump_version,
    get_latest_version,
    get_previous_version,
    get_versions,
)

app = typer.Typer()


@app.command()
def release(repo: Path = typer.Argument(default=Path("."))):
    """
    Create a new release.
    """
    versions = get_versions(repo)
    previous_version = get_latest_version(versions)
    typer.echo(f"Found previous version {previous_version}")

    changes = list(get_commits_between(repo, previous_version))
    typer.echo(f"Found {len(changes)} commits since last release")

    owner, repo_name = parse_remote_url(get_remote_url(repo))

    typer.echo(f"Retrieving pull requests for changes from {owner}/{repo_name}")
    pull_requests = get_pull_requests_for_commits(owner, repo_name, changes)

    labels = set()
    for pull_request in pull_requests:
        labels.update(pull_request.labels)

    config = get_config(repo)
    bump_type = "patch"
    for label in config.major_labels:
        if label in labels:
            typer.echo(f"Detected major version change due label {label}")
            bump_type = "major"
            break

    for label in config.minor_labels:
        if label in labels:
            typer.echo(f"Detected minor version change due label {label}")
            bump_type = "minor"
            break

    if bump_type == "patch":
        typer.echo(
            "Detected patch version change â€” did not see any major or minor labels"
        )

    new_version = bump_version(
        # If there is no previous version, start at 0.0.0
        previous_version or Version("0.0.0"),
        bump_type,
    )
    typer.echo(f"Creating new version {new_version}")

    changelog = generate_changelog(pull_requests, config)
    changelog_file = repo.joinpath("CHANGELOG.md")
    if not changelog_file.exists():
        changelog_file.write_text("# Changelog\n\n")
        typer.echo("Created new changelog file")

    update_existing_changelog = changelog_file.read_text()
    new_changelog = add_or_update_entry(
        new_version, update_existing_changelog, changelog
    ).strip()

    changelog_file.write_text(new_changelog)

    typer.echo("Updated changelog")


@app.command()
def entry(repo: Path = typer.Argument(default=Path(".")), version: str = None):
    """
    Generate a changelog entry for a version.
    If not provided, the current local version from the pyproject.toml file will be used.
    """
    if version is None:
        # Get the version from the pyproject file
        pyproject = tomllib.loads(repo.joinpath("pyproject.toml").read_text())
        version = pyproject["project"]["version"]
        typer.echo(f"Found version {version}")

    # Parse the version
    version = Version(version)

    versions = get_versions(repo)
    previous_version = get_previous_version(versions, version)
    typer.echo(f"Found previous version {previous_version}")

    changes = list(get_commits_between(repo, previous_version))
    typer.echo(f"Found {len(changes)} commits since {previous_version}")

    owner, repo_name = parse_remote_url(get_remote_url(repo))

    typer.echo(f"Retrieving pull requests for changes from {owner}/{repo_name}")
    pull_requests = get_pull_requests_for_commits(owner, repo_name, changes)

    config = get_config(repo)

    changelog = generate_changelog(pull_requests, config)
    print(changelog)


@app.command()
def update(
    repo: Path = typer.Argument(default=Path(".")),
    update_existing: bool = False,
    oldest: str = None,
):
    """
    Update the changelog with all of the releases in the repository.
    Releases are discovered by tags parsable as versions.
    """
    tag_versions = get_versions(repo)
    changelog_file = repo.joinpath("CHANGELOG.md")
    if not changelog_file.exists():
        changelog_file.write_text("# Changelog\n\n")
        typer.echo("Created new changelog file")

    changelog = changelog_file.read_text()
    changelog_versions = get_versions_from_changelog(changelog)

    config = get_config(repo)
    owner, repo_name = parse_remote_url(get_remote_url(repo))

    for version in sorted(tag_versions, reverse=True):
        if oldest is not None and Version(oldest) > version:
            typer.echo(f"Reached oldest version {oldest}, stopping")
            break

        if version in changelog_versions and not update_existing:
            typer.echo(f"Skipping existing version {version}")
            continue
        commits = list(
            get_commits_between(
                repo,
                get_previous_version(tag_versions, version),
                version,
            )
        )

        typer.echo(
            f"Generating changelog entry for {version} with {len(commits)} changes"
        )

        pull_requests = get_pull_requests_for_commits(owner, repo_name, commits)

        changelog = add_or_update_entry(
            version, changelog, generate_changelog(pull_requests, config)
        )

        # Write each entry as we go because I'm developing and impatient :)
        changelog_file.write_text(changelog)

    typer.echo("Updated changelog!")


@app.command()
def sync(
    repo: Path = typer.Argument(default=Path(".")),
    update_existing: bool = False,
    check: bool = False,
    oldest: str = None,
):
    """
    Update releases on GitHub with the local changelog entries.
    """
    changelog_file = repo.joinpath("CHANGELOG.md")
    if not changelog_file.exists():
        typer.echo(
            "Changelog file does not exist. Consider using `rooster backfill` first."
        )
        return

    changelog = changelog_file.read_text()
    changelog_versions = get_versions_from_changelog(changelog)

    owner, repo_name = parse_remote_url(get_remote_url(repo))

    changed = 0
    for version in sorted(changelog_versions, reverse=True):
        if oldest is not None and Version(oldest) > version:
            typer.echo(f"Reached oldest version {oldest}, stopping")
            break

        release = get_release(owner, repo_name, f"v{version}")
        if not release:
            typer.echo(f"Skipping release {version} not found on GitHub")
            continue

        local_entry = entry_to_standalone(extract_entry(changelog, version), version)
        if release.body == local_entry:
            typer.echo(f"Release {version} is up to date")
            continue

        if release.body and not update_existing:
            typer.echo(f"Skipping release {version} with existing body")
            continue

        changed += 1

        if check:
            typer.echo(f"Would update release {version} ")
            continue

        typer.echo(f"Updating release {version}")
        print(local_entry)

    if check:
        typer.echo(f"Would have updated {changed} releases.")
        if changed > 0:
            raise typer.Exit(1)
    else:
        typer.echo(f"Updated {changed} releases.")
