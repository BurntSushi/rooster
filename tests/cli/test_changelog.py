import io
import subprocess
from pathlib import Path

from rooster._pyproject import update_pyproject_version
from rooster._testing import (
    create_tags,
    empty_commit,
    git_directory,
    mock_project,
    mock_pyproject,
    rooster_command,
)


def test_changelog_no_version_no_pyproject(mock_project: Path, snapshot):
    (mock_project / "pyproject.toml").unlink()
    assert rooster_command(["changelog", str(mock_project.resolve())]) == snapshot


def test_changelog_no_version(mock_project: Path, mock_pyproject: Path, snapshot):
    # Should use version from pyproject file
    update_pyproject_version(mock_pyproject, "0.2.0")
    assert rooster_command(["changelog", str(mock_project.resolve())]) == snapshot


def test_changelog_no_remote(mock_project: Path, snapshot):
    assert (
        rooster_command(
            ["changelog", str(mock_project.resolve()), "--version", "0.1.0"]
        )
        == snapshot
    )
