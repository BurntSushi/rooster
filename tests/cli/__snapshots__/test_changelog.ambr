# serializer version: 1
# name: test_changelog_no_remote
  dict({
    'exit_code': 1,
    'stderr': '',
    'stdout': '''
      Found 8 commits
      No remote found; cannot retrieve pull requests to generate changelog entry
  
    ''',
  })
# ---
# name: test_changelog_no_version
  dict({
    'exit_code': 1,
    'stderr': '',
    'stdout': '''
      Found version 0.2.0
      Found 8 commits
      No remote found; cannot retrieve pull requests to generate changelog entry
  
    ''',
  })
# ---
# name: test_changelog_no_version_no_pyproject
  dict({
    'exit_code': 1,
    'stderr': '',
    'stdout': '''
      No pyproject.toml file found; provide a version to generate an entry for.
  
    ''',
  })
# ---
